src/Binance/Entry.cpp

```diff
@@ -29,27 +30,28 @@ string Entry::signJSON(TWCoinType coin, const std::string& json, const Data& key
     return Signer::signJSON(json, key);
 }
 
-HashPubkeyList Entry::preImageHashes(TWCoinType coin, const Data& txInputData) const {
-    auto input = Proto::SigningInput();
-    Data preImageHash;
-    if (input.ParseFromArray(txInputData.data(), (int)txInputData.size())) {
-        const auto signer = Signer(input);
-        // return preimage hash and dummy pubkeyhash (only one signature anyways)
-        preImageHash = signer.preImageHash();
-    }
-    return HashPubkeyList{std::make_pair(preImageHash, Data())};
+Data Entry::preImageHashes(TWCoinType coin, const Data& txInputData) const {
+    return txCompilerTemplate<Proto::SigningInput, TxCompiler::Proto::PreSigningOutput>(
+        txInputData, [](const auto& input, auto& output) {
+            Signer signer(input);
+
+            auto preImageHash = signer.preImageHash();
+            auto preImage = signer.signaturePreimage();
+            output.set_datahash(preImageHash.data(), preImageHash.size());
+            output.set_data(preImage.data(), preImage.size());
```

Comment:
Why is preImage returned here, isn't hash enough?

Reply:
some coins do not need hash, such as solana which just needs preImage. In addition, we will get the preImage to check what contents in it.